tests/test_gates_abstract.py

```diff
@@ -420,6 +423,55 @@ def test_decompose():
     assert isinstance(decomp_gates[0], gates.H)
 
 
+def test_decompose_controlled():
+    backend = NumpyBackend()
+    target = gates.H(0).controlled_by(1)
+    decomp = target.decompose()
+    assert len(decomp) == 1
+    assert isinstance(decomp[0], gates.H)
+    assert decomp[0].control_qubits == (1,)
+    c1 = Circuit(2)
+    c2 = c1.copy()
+    c1.add(target)
+    c2.add(decomp)
+    s1 = backend.execute_circuit(c1).state()
+    s2 = backend.execute_circuit(c2).state()
+    backend.assert_allclose(s1, s2)
+
+
+def test_decompose_controlled_optimized():
+    backend = NumpyBackend()
+    target = gates.RBS(1, 2, 0.1).controlled_by(0)
+    decomp = target.decompose()
+    assert len(decomp) == 6
+    controls_on_zero = sum([0 in g.control_qubits for g in decomp])
+    assert controls_on_zero == 2
+    c1 = Circuit(3)
+    c2 = c1.copy()
+    c1.add(target)
+    c2.add(decomp)
+    s1 = backend.execute_circuit(c1).state()
+    s2 = backend.execute_circuit(c2).state()
+    backend.assert_allclose(s1, s2)
+
+
+@pytest.mark.parametrize(
+    "g1, g2, expected",
+    [
+        (gates.RX(0, pi), gates.RX(0, -pi), True),
+        (gates.RY(1, 0.5), gates.RY(1, -0.5), True),
+        (gates.RZ(2, 2 * pi), gates.RZ(2, -2 * pi), True),
+        (gates.X(0), gates.X(0), True),
+        (gates.Y(1), gates.Y(1), True),
+        (gates.Z(2), gates.Z(2), True),
+        (gates.RX(0, pi / 2), gates.RX(0, pi / 3), False),
```

Comment:
```suggestion
        (gates.X(0), gates.Y(0), False),
        (gates.CNOT(0,1), gates.CNOT(0,1), True),
        (gates.CNOT(0,2), gates.CNOT(1,2), False),
        (gates.RX(0, pi), gates.RX(0, -pi), True),
        (gates.RY(1, 0.5), gates.RY(1, -0.5), True),
        (gates.RZ(2, 2 * pi), gates.RZ(2, -2 * pi), True),
        (gates.X(0), gates.X(0), True),
        (gates.Y(1), gates.Y(1), True),
        (gates.Z(2), gates.Z(2), True),
        (gates.RX(0, pi / 2), gates.RX(0, pi / 3), False),
```
these further cases should help with the coverage